### PR TITLE
Add test for RecursiveUnmount when a submount fails to unmount.

### PR DESCRIPTION
**Pull request by manugupt1**
Add test for RecursiveUnmount when a submount fails to unmount.

https://github.com/moby/sys/issues/45

Signed-off-by: Manu Gupta <manugupt1@gmail.com>

### CODE REVIEW

File: mount/mount_unix.go
Comment (kolyshkin) [inline mount/mount_unix.go:24]: I'd rather not do that for the sake of a single unit test.

Unfortunately I don't know of any other way to make the deferred unmount fail.

Comment (kolyshkin) [reply mount/mount_unix.go:24]: Hmm, maybe I do.

Try to create a mount which is shadowed by another mount. This will be shown in mountinfo, but the directory will not be there, so the unmount will fail with ENOENT.

IOW something like
1. mount "tmp/dir1/subdir" using tmpfs.
2. mount "tmp/dir1" using tmpfs.

Now, RecursiveUnmount will try to unmount "tmp/dir1/subdir", and since it's hidden by the second mount, its unmount will fail with ENOENT.

File: mount/mount_unix_test.go
Comment (kolyshkin) [inline mount/mount_unix_test.go:281]: I see that you copied the "TooGreedy" test except for its sanity check part. Do you think it's not necessary?

File: mount/mount_unix.go
Comment (manugupt1) [reply mount/mount_unix.go]: Thanks for the suggestion. I did not know about shadow mounts. I updated the test by using your suggestion.

**Comment by manugupt1** [comment]
HI @kolyshkin 
Will you be able to review this when you get a chance?

Thanks

**Comment by kolyshkin** [comment]
LGTM, except I don't understand why 3 mounts are needed -- it seems that 2 are enough.

One other thing, can you please add a short description of what the added case tests (something like "$NAME checks that RecursiveUnmount does return an error when a submount fails to be unmounted")

**Comment by manugupt1** [comment]
> LGTM, except I don't understand why 3 mounts are needed -- it seems that 2 are enough.
> 
> One other thing, can you please add a short description of what the added case tests (something like "$NAME checks that RecursiveUnmount does return an error when a submount fails to be unmounted")

Having 3 mounts ensures that we check for the cases when suberr is nil and not nil.

File: mount/mount_unix_test.go
Comment (kolyshkin) [inline mount/mount_unix_test.go:269]: This comment no longer reflects reality -- I don't see "dir" and "dir-other" in here.

File: mount/mount_unix_test.go
Comment (kolyshkin) [inline mount/mount_unix_test.go:255]: I guess it makes sense to describe the idea here, so that the future readers will get a grasp about what is going on.

```go 
// Create a set of mounts that should result in RecursiveUnmount failure,
// caused by the fact that the grandchild mount is shadowed by the child mount,
// and the child mount is shadowed by the parent mount. So. these two mounts
// are listed in mountinfo, but since they are unreachable, unmount will fail.
```

File: mount/mount_unix_test.go
Comment (thaJeztah) [inline mount/mount_unix_test.go:16]: Minor nit; can you remove the empty line and sort the imports?

**Comment by thaJeztah** [comment]
@manugupt1 with https://github.com/moby/sys/pull/109 merged, do you need new releases of `mount` and `mountinfo` after this was merged, or do you anticipate other changes coming before we should do a new release?

**Comment by kolyshkin** [comment]
> I do not know enough k8s to see if I can base on the latest commit over master and then do follow ups either here or there if reqd.

Technically you can, but maintainers will complain. For some package which did not do a release for a long time they can make an exception, and of course you can use an untagged version for testing.

So, let us know once you finished the testing and need a release, and we'll tag one.

File: mount/mount_unix_test.go
Comment (kolyshkin) [inline mount/mount_unix_test.go:271]: So, with this t.Run we have two subtests which are similarly named. I guess either to remove t.Run, or make the subtest name different (e.g. by `t.Run(shadowedMount)`).

@thaJeztah WDYT?

File: mount/mount_unix_test.go
Comment (thaJeztah) [inline mount/mount_unix_test.go:265]: Didn't leave it as a comment earlier as it's a minor nit, but if you're updating, perhaps put the `//nolint` after the line; slightly cleaner, and makes it slightly clearer which line the `//nolint` applies to.

```go
defer Unmount(dir) //nolint:errcheck
```

File: mount/mount_unix_test.go
Comment (thaJeztah) [reply mount/mount_unix_test.go:271]: Good catch; yes, I think that would work; I'd change this to:

```go
		t.Run(shadowedMount, func(t *testing.T) {
```

File: mount/mount_unix_test.go
Comment (thaJeztah) [inline mount/mount_unix_test.go:276]: Given that `errors.Is()` already returns `false` if `err == nil`, this can be simplified to:

```suggestion
			err := RecursiveUnmount(shadowedMount)
			if !errors.Is(err, unix.ENOENT) {
				t.Fatalf("expected submount(shadowed) %s to return unix.ENOENT, got %v", shadowedMount, err)
			}
```

File: mount/mount_unix_test.go
Comment (thaJeztah) [inline mount/mount_unix_test.go:252]: This seems making things more complicated than needed to find paths we already know what they'll be (relative to `tmp` that is). Given that this only runs on unix/linux, we don't strictly need `filepath` (path separator will always be `/`) and because these paths are "fixed", we don't need to care about path-escaping etc. (also `grandchild` should probably be `grandChild` to be properly camelCase); 

perhaps something like:

```go
var (
	tmp        = t.TempDir()
	parent     = tmp + "/sub1"
	child      = tmp + "/sub1/sub2"
	grandChild = tmp + "/sub1/sub2/sub3"
)

err := os.MkdirAll(grandChild, 0o700)
if err != nil {
	t.Fatal(err)
}
```

That way, it's obvious at a glance what paths we're dealing with (the `var(    )` block makes them align nicely, so it's easier to read for this).

**Comment by thaJeztah** [comment]
Thanks for updating; changes look good - could you squash the commits? (it's fine to squash my suggestions with your commit 👍)

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

@kolyshkin PTAL

**Review by kolyshkin — APPROVED** [review]
LGTM

**Pull request merged**
